### PR TITLE
MLDB-1618 fix two very obvious bugs in ridge regression and allow cla…

### PR DESCRIPTION
**Pull request by steadtler**
…ssifier to pass the regularization factor instead of always doing the auto-determination.

The two bugs were introduced long ago in a performance refactoring, and would result in always using the default parameter.

### CODE REVIEW

File: ml/algebra/least_squares.cc [hunk 1]
```diff
@@ -284,125 +285,124 @@ struct RidgeRegressionIteration {
         int m = A.shape()[0];
         int n = A.shape()[1];
 
-            double current_lambda = lambda;
```
Comment (steadtler) [inline ml/algebra/least_squares.cc]: 1st bug: lambda & current_lambda cant be the same.

File: ml/algebra/least_squares.cc [hunk 2]
```diff
@@ -284,125 +285,124 @@ struct RidgeRegressionIteration {
...
+        distribution<Float> A_A_pinv_diag(m);
+        for (unsigned j = 0;  j < m;  ++j)
+            A_A_pinv_diag[j] = SIMD::vec_dotprod_dp(&A[j][0], &A_pinv[j][0], n);
 #endif
 
-            doneStep("A_A_pinv");
+        doneStep("A_A_pinv");
 
-            // Now figure out the performance
-            double total_mse_biased = 0.0, total_mse_unbiased = 0.0;
```
Comment (steadtler) [inline ml/algebra/least_squares.cc]: 2nd bug: local variable shadowing a member.

File: ml/algebra/irls.cc [hunk 3]
```diff
@@ -353,7 +353,8 @@ perform_irls_unconditioned(const distribution<Float> & correct,
                            const boost::multi_array<Float, 2> & outputs,
                            const distribution<Float> & w,
                            Link_Function link_function,
-                           bool ridge_regression)
+                           bool ridge_regression,
+                           Float regularization_factor)
```
Comment (guyd) [inline ml/algebra/irls.cc]: Consider changing the Float parameter to an optional<Float>.

File: ml/algebra/irls.cc [hunk 4]
```diff
@@ -398,7 +399,8 @@ perform_irls_conditioned(const distribution<Float> & correct,
                               const boost::multi_array<Float, 2> & outputs,
                               const distribution<Float> & w,
                               Link_Function link_function,
-                              bool ridge_regression)
+                              bool ridge_regression, 
+                              Float regularization_factor)
```
Comment (guyd) [inline ml/algebra/irls.cc]: Same here optional<Float>

File: ml/algebra/irls.cc [hunk 5]
```diff
@@ -597,14 +601,15 @@ perform_irls(const distribution<double> & correct,
              const distribution<double> & w,
              Link_Function link_function,
              bool ridge_regression,
+             double regularization_factor,
```
Comment (guyd) [inline ml/algebra/irls.cc]: and here.

File: ml/algebra/irls.h [hunk 6]
```diff
@@ -298,6 +298,7 @@ perform_irls(const distribution<float> & correct,
              const distribution<float> & w,
              Link_Function link_function,
              bool ridge_regression = true,
+             float regularization_factor = 1e-5,
```
Comment (guyd) [inline ml/algebra/irls.h]: I guess the optional would not work if we want to have a default factor.

File: ml/algebra/testing/least_squares_test.cc [hunk 7]
```diff
@@ -158,7 +158,75 @@ void do_test2()
...
+
+    Float tol = 20; /* percent */
+
+    BOOST_REQUIRE_EQUAL(x.size(), 4);
+
+    distribution<Float> error1 = (A * x) - b;
+    distribution<Float> error2 = (A * x2) - b;
+
+
+    BOOST_CHECK_CLOSE(x2[0], (Float)0.0034078, tol);
```
Comment (guyd) [inline ml/algebra/testing/least_squares_test.cc]: Don't you mean `BOOST_CHECK_CLOSE(x2[0], x1[0], tol)`?

File: ml/algebra/testing/least_squares_test.cc [hunk 8]
```diff
@@ -158,7 +158,75 @@ void do_test2()
...
+    BOOST_CHECK_CLOSE(x2[0], (Float)0.0034078, tol);
```
Comment (guyd) [inline ml/algebra/testing/least_squares_test.cc]: Duplicate

File: ml/algebra/testing/least_squares_test.cc [hunk 9]
```diff
@@ -158,7 +158,75 @@ void do_test2()
...
+
+    if (x2[1] == 0.0)
+        BOOST_CHECK_CLOSE(x2[1], (Float)0.0000000, tol);
```
Comment (guyd) [inline ml/algebra/testing/least_squares_test.cc]: ?

File: ml/algebra/testing/least_squares_test.cc [hunk 10]
```diff
@@ -158,7 +158,75 @@ void do_test2()
...
+    else BOOST_CHECK(abs(x2[1]) < 1e-10);
+
+    if (x2[2] == 0.0)
```
Comment (guyd) [inline ml/algebra/testing/least_squares_test.cc]: ?

File: ml/algebra/testing/least_squares_test.cc [hunk 11]
```diff
@@ -158,7 +158,75 @@ void do_test2()
...
+        BOOST_CHECK_CLOSE(x2[2], (Float)0.0000000, tol);
+    else BOOST_CHECK(abs(x2[2]) < 1e-10);
+    
+    BOOST_CHECK_CLOSE(x2[3], (Float)0.0000003, tol);
```
Comment (guyd) [inline ml/algebra/testing/least_squares_test.cc]: close to `x1[3]`?

File: ml/jml/glz_classifier_generator.cc [hunk 12]
```diff
@@ -82,6 +84,8 @@ options() const
              "which link function to use for the output function")
         .add("ridge_regression", ridge_regression,
              "use a regularized regression with smaller coefficients")
+        .add("regularization_factor", regularization_factor, "-1 to infinite",
+             "regularization factor to use. Auto-determined if negative. Auto-determined is slower.")
```
Comment (guyd) [inline ml/jml/glz_classifier_generator.cc]: Add the default value as the last parameter.

**Comment by guyd** [comment]
+1  except for the confusion in the test.

**Comment by nicolaskruchten** [comment]
Is this on track to be merged?

**Comment by mailletf** [comment]
@steadtler: fermé mais pas mergé??

**Pull request closed without merging**